File: swift/Sources/TokenTrackerCore/UpdateChecker.swift
//
//  UpdateChecker.swift
//  TokenTrackerCore
//
//  移植 tokentracker/updatecheck.py：GitHub Releases 更新检查，缓存 24h，
//  网络失败静默，绝不阻塞启动。
//

import Foundation

public struct UpdateInfo: Equatable, Sendable {
    public var latest: String
    public var url: String
    public var checkedAt: Double

    public init(latest: String, url: String, checkedAt: Double) {
        self.latest = latest
        self.url = url
        self.checkedAt = checkedAt
    }

    enum CodingKeys: String, CodingKey {
        case latest, url
        case checkedAt = "checked_at"
    }
}

extension UpdateInfo: Codable {}

public struct UpdateChecker: Sendable {
    public static let repo = "fengfe1125/tokentracker"
    public static let cacheTTL: Double = 24 * 3600

    public let cachePath: String
    public var clock: @Sendable () -> Double
    /// 注入缝：网络抓取（测试注入）。输入 URL，返回响应 Data。
    public var fetch: @Sendable (String) throws -> Data

    public init(cachePath: String? = nil, clock: (@Sendable () -> Double)? = nil,
                fetch: (@Sendable (String) throws -> Data)? = nil) {
        self.cachePath = cachePath ?? NSHomeDirectory() + "/.tokentracker/update_check.json"
        self.clock = clock ?? { Date().timeIntervalSince1970 }
        self.fetch = fetch ?? { url in
            var request = URLRequest(url: URL(string: url)!)
            request.setValue("TokenTracker-update-check", forHTTPHeaderField: "User-Agent")
            request.timeoutInterval = 5
            final class ResultBox: @unchecked Sendable {
                var result: Result<Data, Error> = .failure(SQLiteError(message: "no response"))
            }
            let box = ResultBox()
            let semaphore = DispatchSemaphore(value: 0)
            URLSession.shared.dataTask(with: request) { data, _, error in
                if let data { box.result = .success(data) }
                else { box.result = .failure(error ?? SQLiteError(message: "unknown")) }
                semaphore.signal()
            }.resume()
            _ = semaphore.wait(timeout: .now() + 8)
            return try box.result.get()
        }
    }

    public static func parseVersion(_ v: String) -> [Int] {
        v.replacingOccurrences(of: "^v", with: "", options: .regularExpression)
            .split(separator: ".")
            .compactMap { $0.allSatisfy(\.isNumber) ? Int($0) : nil }
    }

    public func readCache() -> UpdateInfo? {
        guard let data = FileManager.default.contents(atPath: cachePath),
              let info = try? JSONDecoder().decode(UpdateInfo.self, from: data),
              !info.latest.isEmpty else { return nil }
        return info
    }

    private func writeCache(_ info: UpdateInfo) {
        try? FileManager.default.createDirectory(
            atPath: (cachePath as NSString).deletingLastPathComponent,
            withIntermediateDirectories: true)
        if let data = try? JSONEncoder().encode(info) {
            FileManager.default.createFile(atPath: cachePath, contents: data)
        }
    }

    /// 返回最新发布信息或 nil（离线/无 release 时给过期缓存或 nil）。
    @discardableResult
    public func check(force: Bool = false) -> UpdateInfo? {
        let cached = readCache()
        if !force, let cached, clock() - cached.checkedAt < Self.cacheTTL {
            return cached
        }
        do {
            let data = try fetch("https://api.github.com/repos/\(Self.repo)/releases/latest")
            let obj = try JSONSerialization.jsonObject(with: data) as? [String: Any] ?? [:]
            let info = UpdateInfo(latest: obj["tag_name"] as? String ?? "",
                                  url: obj["html_url"] as? String ?? "",
                                  checkedAt: clock())
            writeCache(info)
            return info
        } catch {
            return cached   // 失败静默：用过期缓存或 nil
        }
    }

    public static func updateAvailable(_ info: UpdateInfo?, current: String) -> Bool {
        guard let info, !info.latest.isEmpty else { return false }
        // Python tuple 比较：逐元素，短序列前缀相等时更长者大
        let latest = parseVersion(info.latest)
        let currentParts = parseVersion(current)
        for index in 0..<max(latest.count, currentParts.count) {
            let l = index < latest.count ? latest[index] : 0
            let c = index < currentParts.count ? currentParts[index] : 0
            if l != c { return l > c }
        }
        return false
    }
}

File: windows/src/TokenTracker.Core/Updates/UpdateChecker.cs
using System.Text.RegularExpressions;
using TokenTracker.Core.Json;
using TokenTracker.Core.Platform;

namespace TokenTracker.Core.Updates;

public sealed record UpdateInfo(string Latest, string Url, double CheckedAt);

/// <summary>
/// GitHub Releases 更新检查（移植 UpdateChecker.swift）：缓存 24h（与 macOS 共用 update_check.json），
/// 网络失败静默。第一期只提示并打开发布页，不做应用内安装。
/// </summary>
public sealed class UpdateChecker
{
    public const string Repo = "fengfe1125/tokentracker";
    public const double CacheTtl = 24 * 3600;
    public static string ReleasesUrl => $"https://github.com/{Repo}/releases/latest";

    static readonly HttpClient Client = new() { Timeout = TimeSpan.FromSeconds(8) };

    public string CachePath { get; init; } = Path.Combine(WinPaths.DataDir, "update_check.json");
    public Func<double> Clock { get; init; } = () => DateTimeOffset.UtcNow.ToUnixTimeMilliseconds() / 1000.0;

    /// <summary>注入缝：网络抓取。</summary>
    public Func<string, byte[]> Fetch { get; init; } = url =>
    {
        using var request = new HttpRequestMessage(HttpMethod.Get, url);
        request.Headers.TryAddWithoutValidation("User-Agent", "TokenTracker-update-check");
        using var response = Client.Send(request);
        response.EnsureSuccessStatusCode();
        return response.Content.ReadAsByteArrayAsync().GetAwaiter().GetResult();
    };

    public static int[] ParseVersion(string v) =>
        Regex.Replace(v, "^v", "").Split('.')
            .Where(p => p.Length > 0 && p.All(char.IsAsciiDigit)).Select(int.Parse).ToArray();

    public UpdateInfo? ReadCache()
    {
        var obj = PyJson.ReadObjectFile(CachePath);
        return obj?.Get("latest") is string { Length: > 0 } latest
            ? new UpdateInfo(latest, obj.Get("url") as string ?? "", PyJson.AsNumber(obj.Get("checked_at")) ?? 0)
            : null;
    }

    /// <summary>返回最新发布信息或 null（离线/无 release 时给过期缓存或 null）。</summary>
    public UpdateInfo? Check(bool force = false)
    {
        var cached = ReadCache();
        if (!force && cached is not null && Clock() - cached.CheckedAt < CacheTtl) return cached;
        try
        {
            var obj = PyJson.ParseObject(Fetch($"https://api.github.com/repos/{Repo}/releases/latest")) ?? [];
            var info = new UpdateInfo(obj.Get("tag_name") as string ?? "", obj.Get("html_url") as string ?? "", Clock());
            SharedFile.AtomicWriteText(CachePath, PyJson.Serialize(new Dictionary<string, object?>
            {
                ["latest"] = info.Latest, ["url"] = info.Url, ["checked_at"] = info.CheckedAt,
            }));
            return info;
        }
        catch (Exception e) when (e is HttpRequestException or TaskCanceledException or IOException)
        {
            return cached; // 失败静默
        }
    }

    public static bool UpdateAvailable(UpdateInfo? info, string current)
    {
        if (info is null || info.Latest.Length == 0) return false;
        // Python tuple 比较：逐元素，短序列前缀相等时更长者大
        var latest = ParseVersion(info.Latest);
        var currentParts = ParseVersion(current);
        for (var i = 0; i < Math.Max(latest.Length, currentParts.Length); i++)
        {
            var l = i < latest.Length ? latest[i] : 0;
            var c = i < currentParts.Length ? currentParts[i] : 0;
            if (l != c) return l > c;
        }
        return false;
    }
}
